Time scale improvements

tredston opened the pull request:
Complete work started in https://github.com/chartjs/Chart.js/tree/time-scale-improvements (#3673) to convert `scale.time` to internally use unix timestamps instead of moments. Improves performance and reduces complexity.

etimberg commented:
Thanks @tredston I will look at reviewing this. In the meantime, could you rebase against master and squash into some smaller commits?

etimberg commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+		var stepSize = generationOptions.stepSize;
+		ticks.push(generationOptions.min !== undefined ? generationOptions.min : niceRange.min);
+		var cur = moment(niceRange.min);
+		while (cur.add(stepSize, generationOptions.unit).valueOf() < niceRange.max) {
```
Does this still need to use moment? I thought once everything was timestamps this wasn't needed? Or does this make things a lot simpler overall?

tredston replied:
This simplifies units which don't have a constant number of milliseconds in them eg months, years.

etimberg replied:
ok, makes sense 😄

etimberg commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+		var niceMin;
+		var niceMax;
+		var isoWeekday = generationOptions.isoWeekday;
+		if (isoWeekday !== false) {
```
I think this option only applies if `generationOptions.unit` is `'week'` otherwise it could be transforming data that is in seconds, or minutes.

tredston replied:
Yep, will update to only apply when `generationOptions.unit` is `week`.

tredston replied:
Updated.

etimberg commented on src/scales/scale.time.js:
```diff
@@ -339,71 +382,78 @@ module.exports = function(Chart) {
+		getLabelCapacity: function(exampleTime) {
+			var me = this;
+
+			me.displayFormat = me.options.time.displayFormats.millisecond;	// Pick the longest format for guestimation
```
would be nice if we didn't have to set this on ourself .. could we refactor to pass this into `tickFormatFunction` instead?

tredston replied:
The problem is that `convertTicksToLabels` uses `tickFormatFunction` as the callback for a `map` so we can't really change `tickFormatFunction`'s message signature to include the display format.

etimberg commented on test/scale.time.tests.js:
```diff
@@ -245,111 +220,131 @@ describe('Time scale tests', function() {
-		// scale.buildTicks();
-		scale.update(400, 50);
+		var scale = createScale(mockData, config);
+		scale.update(5000, 200);
```
does this need to be this wide?

tredston replied:
It just needed to be wide enough to not be in danger of having ticks dropped for overlapping. Looking again this can be revised down to about `2500` before the last two ticks might overlap. 

Alternatively we can have a narrower graph and update the `expect` to remove the penultimate tick.

etimberg commented on test/scale.time.tests.js:
```diff
@@ -245,111 +220,131 @@ describe('Time scale tests', function() {
-		// scale.buildTicks();
-		scale.update(400, 50);
+		var scale = createScale(mockData, config);
+		scale.update(800, 200);
```
same comment here about the width

tredston replied:
As above

etimberg commented on test/scale.time.tests.js:
```diff
@@ -245,111 +220,131 @@ describe('Time scale tests', function() {
 
-	it('Should use the isoWeekday option', function() {
-		var scaleID = 'myScale';
+	it('should not have overlapping ticks', function() {
```
Any reason there are two tests with the same name?

tredston replied:
One (in `describe('when specifying limits'...` suite) covers the case when specifying a `time.max` while the other covers the general case. Since they are in separate suites jasmine will print different messages if/when they break.

etimberg replied:
ok, just wanted to make sure this wasn't an error 😄

etimberg commented on test/scale.time.tests.js:
```diff
@@ -245,111 +220,131 @@ describe('Time scale tests', function() {
+			}
+		});
+		var scale = chart.scales.xScale0;
+		scale.update(24633, 160);
```
this seems like a giant width :/

tredston replied:
This is essentially a regression test, the code was lifted from the example in #2249.

etimberg replied:
ah, ok. that makes sense

etimberg commented on test/scale.time.tests.js:
```diff
@@ -395,29 +378,94 @@ describe('Time scale tests', function() {
-		expect(xScale.getPixelForValue('', 0, 0)).toBeCloseToPixel(71);
-		expect(xScale.getPixelForValue('', 6, 0)).toBeCloseToPixel(452);
-		expect(xScale.getPixelForValue('2015-01-01T20:00:00')).toBeCloseToPixel(71);
+		it('should be bounded by midnights', function() {
```
[nit] Is the goal of this test to ensure that that when days are displayed, the last tick is at the end of the last day and similarly for the first tick?

After reading the several years ones, I think a better title might be "should be bounded by the nearest day beginnings" since I was a little unclear on what this meant at first.

tredston replied:
Yes that's the goal. 

That title is much clearer, I will update.

tredston replied:
Updated.

tredston commented on src/scales/scale.time.js:
```diff
@@ -8,34 +8,42 @@ module.exports = function(Chart) {
-			name: 'month',
+		},
+		month: {
+			size: 2.628e9,
```
These sizes are used to estimate the unit and step size to use, they aren't used for calculating the tick locations. The step size is a multiple of the unit, so a 'unit' of 'month' and 'stepSize' of '3' give ticks which are 3 months apart. The labels are still rounded (by `moment`) to the start of the `unit` so they display correctly.

This very issue is addressed in https://github.com/chartjs/Chart.js/pull/3914/commits/79230683d19f860c6bd2ffc61429b583122c53fb

etimberg replied:
Ah, ok, great 👍

tredston commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+		var units = Object.keys(time);
+		var unit;
+
+		for (var i = units.indexOf(minUnit); i < units.length; i++) {
```
I prefer to use a for loop when doing simple array iteration, I think having the start index, end conditions and stride all together is clearer than spreading them out. 

The `fits` variable would also need to be hoisted out of the loop, in which case we could write:
```javascript
for (var i = units.indexOf(minUnit); i < units.length && !fits; i++)
``` 
instead?

etimberg replied:
i think we could leave this as is. I had just wondered if we could make it a bit simpler but it's already straightforward

tredston replied:
OK 👍

tredston replied:
I've found and applied a further simplification to this section of code.

etimberg commented:
Tried these out and they work well. It seems really fast now 😄 I did notice that the add buttons on https://github.com/chartjs/Chart.js/blob/master/samples/scales/time/line-time-point-data.html don't work. This was likely broken when I started on this not by your changes afterwards. Do you mind fixing it while you're at it?

tredston commented:
Squashed to fewer commits and rebased onto master. 

Updated the code following discussions thus far.

Updated [the sample](https://github.com/chartjs/Chart.js/blob/master/samples/scales/time/line-time-point-data.html) as requested.

tredston commented on samples/scales/time/line-time-point-data.html:
```diff
@@ -119,7 +119,7 @@
 		document.getElementById('addData').addEventListener('click', function() {
 			if (config.data.datasets.length > 0) {
-				var lastTime = myLine.scales['x-axis-0'].labelMoments[0].length ? myLine.scales['x-axis-0'].labelMoments[0][myLine.scales['x-axis-0'].labelMoments[0].length - 1] : moment();
+				var lastTime = myLine.scales['x-axis-0'].ticksAsTimestamps.length ? moment(myLine.scales['x-axis-0'].ticksAsTimestamps[myLine.scales['x-axis-0'].ticksAsTimestamps.length - 1]) : moment();
```
Done.

tredston commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+		} else if (label.isValid && label.isValid()) {
+			// Moment support
+			return label;
+		} else if (typeof timeOpts.format !== 'string' && timeOpts.format.call) {
```
Done.

tredston commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+			return label;
+		} else if (typeof timeOpts.format !== 'string' && timeOpts.format.call) {
+			// Custom parsing (return an instance of moment)
+			console.warn('options.time.format is deprecated and replaced by options.time.parser. See http://nnnick.github.io/Chart.js/docs-v2/#scales-time-scale');
```
Done.

tredston commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+			// Custom parsing (return an instance of moment)
+			console.warn('options.time.format is deprecated and replaced by options.time.parser. See http://nnnick.github.io/Chart.js/docs-v2/#scales-time-scale');
+			return timeOpts.format(label);
+		}
```
Done.

tredston commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+
+		if (unitDefinition.steps) {
+			// Have an array of steps
+			for (var i = 0; i < unitDefinition.steps.length && sizeInUnits > maxTicks; i++) {
```
Done.

tredston commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+			ticks.push(cur.valueOf());
+		}
+		var realMax = generationOptions.max || niceRange.max;
+		var minSpacing = (dataRange.max - dataRange.min) / generationOptions.maxTicks;
```
Fixed.

tredston commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+				// Auto determine step size
+				stepSize = determineStepSize(minTimestamp || dataMin, maxTimestamp || dataMax, unit, maxTicks);
 			}
+			var timeGeneratorOptions = {
```
Done.

tredston commented on src/scales/scale.time.js:
```diff
@@ -339,71 +382,78 @@ module.exports = function(Chart) {
+			var me = this;
+			var epochWidth = me.max - me.min;
+			var decimal = 0;
+			if (epochWidth > 0) {
```
Done.

tredston commented on src/scales/scale.time.js:
```diff
@@ -339,71 +382,78 @@ module.exports = function(Chart) {
-			return moment(label, me.options.time.format);
+
+			var tickLabelWidth = me.ctx.measureText(label).width;
+			var cosRotation = Math.cos(helpers.toRadians(me.options.ticks.maxRotation));
```
Done.

tredston commented:
Thanks for the review @simonbrunel.

I've implemented the requested changes around caching local variables and logic simplification.

However I'm not keen to implement:
> gather local variables in function header (one per line, except for uninitialized variables)

beacuse:
1. there are linting rules for this (`one-var` and `vars-on-top`) which are explicitly disabled.
2. splitting up declaration and assignment makes the code more difficult to follow, as well as more difficult to migrate to using `const` when/if ever you adopt ES6.
3. it makes diffing harder, eg from 
```javascript
var i, j;
```
to
```javascript
var i = 0;
var j;
```
4. it only saves 63 bytes when minified, or 0.04% of `Chart.min.js`, which I don't think outweighs the above.

tredston commented on src/scales/scale.time.js:
```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
 				quarter: '[Q]Q - YYYY', // Q3
 				year: 'YYYY' // 2015
-			}
+			},
```
Because there were multiple authors and linting (`comma-dangle`) allows it :)

simonbrunel replied:
Of course, I mean you modified both lines, but only added comma on the second one. Anyway, that's not important!

simonbrunel commented:
Thanks @tredston for the changes,

1. `one-var` and `vars-on-top` are not enforced because we changed linting rules after years of development and it would have required to much change (that's why it's not applied everywhere).

2. that's a point, we still use `var` not `const` and IMO it's safer to have `var` declared at the top of the scope it's introduced.

3. not sure I follow that point.

4. and from a global perspective? I agree that there is room for refactor all around the code, so it may sound useless but actually it makes a small appreciable difference when minified.

5. kind related: IIRC in QML JavaScript engine, declaring a var inside a loop is (was?) a perf killer, so still better to have it declared outside.

Anyway, that's minor, we can still rework that later, too bad that you removed your last commit implementing these changes ;)

hmcfletch commented:
I pulled this branch to use on my own site because I was having the overlapping tick issue with some of my time scales. It seems that instead of picking better start and end values to allow for evenly placed ticks, it simply drops what would have been the second to last tick. The makes for a much larger gap between the last two ticks on the axis when displayed.

It is definitely better than the current situation, but there is still an underlying issue that needs to be resolved.

etimberg commented:
@hmcfletch is it possible to drop a test case here that uses this branch?

hmcfletch commented:
Working on a deploy today, but i'll see what i can do.

etimberg commented:
@hmcfletch were you able to make any progress on a test case for the issue you were seeing?

hmcfletch commented:
Here is the data I am working with:

**Screenshot of Chart**
https://www.evernote.com/shard/s63/sh/f8cf9342-d934-43e8-820d-6bc4d49cf862/9edfdb6a5428fbcb

**Chart Data**
```json
{
   "labels":[
      "2017-02-03T00:00:00.000Z",
      "2017-02-04T00:00:00.000Z",
      "2017-02-06T00:00:00.000Z",
      "2017-02-07T00:00:00.000Z",
      "2017-02-08T00:00:00.000Z",
      "2017-02-09T00:00:00.000Z",
      "2017-02-10T00:00:00.000Z",
      "2017-02-11T00:00:00.000Z",
      "2017-02-13T00:00:00.000Z",
      "2017-02-14T00:00:00.000Z",
      "2017-02-15T00:00:00.000Z",
      "2017-02-16T00:00:00.000Z",
      "2017-02-17T00:00:00.000Z",
      "2017-02-19T00:00:00.000Z",
      "2017-02-20T00:00:00.000Z",
      "2017-02-21T00:00:00.000Z",
      "2017-02-22T00:00:00.000Z",
      "2017-02-23T00:00:00.000Z",
      "2017-02-24T00:00:00.000Z",
      "2017-02-26T00:00:00.000Z",
      "2017-02-27T00:00:00.000Z",
      "2017-02-28T00:00:00.000Z",
      "2017-03-01T00:00:00.000Z"
   ],
   "datasets":[
      {
         "fill":false,
         "borderWidth":2,
         "borderColor":"rgba(145, 226, 254, 1)",
         "pointBackgroundColor":"rgba(37, 198, 254, 1)",
         "pointHoverBackgroundColor":"rgba(0, 237, 164, 1)",
         "data":[
            {
               "x":"2017-02-02",
               "y":168244
            },
            {
               "x":"2017-02-03",
               "y":168306
            },
            {
               "x":"2017-02-05",
               "y":168353
            },
            {
               "x":"2017-02-06",
               "y":168432
            },
            {
               "x":"2017-02-07",
               "y":168476
            },
            {
               "x":"2017-02-08",
               "y":168518
            },
            {
               "x":"2017-02-09",
               "y":168596
            },
            {
               "x":"2017-02-10",
               "y":168650
            },
            {
               "x":"2017-02-12",
               "y":168727
            },
            {
               "x":"2017-02-13",
               "y":168780
            },
            {
               "x":"2017-02-14",
               "y":168820
            },
            {
               "x":"2017-02-15",
               "y":168860
            },
            {
               "x":"2017-02-16",
               "y":168926
            },
            {
               "x":"2017-02-18",
               "y":168951
            },
            {
               "x":"2017-02-19",
               "y":169032
            },
            {
               "x":"2017-02-20",
               "y":169078
            },
            {
               "x":"2017-02-21",
               "y":169125
            },
            {
               "x":"2017-02-22",
               "y":169155
            },
            {
               "x":"2017-02-23",
               "y":169193
            },
            {
               "x":"2017-02-25",
               "y":169277
            },
            {
               "x":"2017-02-26",
               "y":169300
            },
            {
               "x":"2017-02-27",
               "y":169330
            },
            {
               "x":"2017-02-28",
               "y":169344
            }
         ]
      }
   ]
}
```
**Chart Options**
```json
{
  "responsive": true,
  "legend": {
    "display": false
  },
  "tooltips": {
    "callbacks": {},
    "displayColors": false
  },
  "scales": {
    "yAxes": [
      {
        "type": "linear",
        "ticks": {
          "beginAtZero": false,
          "fontColor": "#7b8994"
        }
      }
    ],
    "xAxes": [
      {
        "ticks": {
          "minRotation": 45,
          "fontColor": "#7b8994"
        },
        "type": "time",
        "time": {
          "format": "YYYY-MM-DD",
          "tooltipFormat": "ll HH:mm"
        }
      }
    ]
  }
}
```

etimberg commented:
@tredston @hmcfletch the behaviour you're seeing is happening due to https://github.com/tredston/Chart.js/blob/time-scale-improvements/src/scales/scale.time.js#L185-L189

I can't recall what that was originally designed to achieve. I think it was something to always ensure the last tick is present if the size of the axis is not a multiple of the step size. 

Options I see:
1. leave it as is (regression from current behaviour)
2. Always push `realMax` if it is not the last value in the array (don't want any duplicate tick marks).

I personally prefer option 2. @simonbrunel thoughts? @tredston could you make those changes?

etimberg commented:
I tried changing that if to:
```javascript
if (ticks[ticks.length - 1] !== realMax) {
    ticks.push(realMax);
}
```

and it matched v2.5

tredston commented:
That code was meant to avoid adding the penultimate tick if it was in danger of overlapping with the last tick. Seemingly label rotation was not accounted for. 

Requested change applied and rebased onto master.

etimberg commented:
Thanks @tredston 

@simonbrunel shall we merge this?

simonbrunel commented:
Sure!
